### ISSUE TITLE
Parsons problems, i.e. rearranging shuffled code

### ISSUE THREAD

**Issue by alexmojaki** [alexmojaki/futurecoder#53]
An instruction/assessment method which seems to be very effective is Parsons problems, where the student is given complete code but out of order, and they have to rearrange the lines into the correct order. This is easier and less stressful, and allows practicing a concept even if you haven't mastered all of the details.

Some useful links:

- https://runestone.academy/runestone/static/instructorguide/Directives/mixedUp.html
- https://cseducators.stackexchange.com/questions/302/effectiveness-of-parsons-problems

I can imagine this being used in two ways:

- As exercises in specific steps in the course.
- As a way of 'revealing' the solution to arbitrary exercises. Automatically giving the user a shuffled solution for any exercise is easy. I'm just not sure what the user experience should be. We currently allow the user to reveal the solution bit by bit until they see the whole thing. If we simply replaced this entirely with a Parsons problem, what if the user gets really stuck and can't solve that either? If we add a final button to just show the correctly ordered solution, how do we prevent the temptation to just click that immediately? Is there anything from the existing solution reveal method that's worth keeping? Should we give the user a choice between solution reveal methods?

The simplest way to implement Parsons problems is to just display out of order code and let them copy/type into the editor. We can do that already. This relates a bit to #35.

A fancier solution would be to use https://js-parsons.github.io/ to allow nice dragging and dropping.

I'm keen to hear opinions and ideas, particularly about using Parsons problems to reveal the solution.

**Comment by eddieantonio** [comment]
Parsons problems are SUPER effective, especially in the early stages of learning how to code. A major stumbling block for novice coders is syntax — getting all the parentheses, colons, and everything in the right order ([I should know][sensibility] :joy:).

When teaching about control flow especially, I'd be keen to the solution expressed as Parsons problems first. I have anecdotal experience of students tripping over the colon at the end of if- and for- statements, and unsure about the role of indentation. I think introducing these concepts with Parsons problems makes it easier to focus on logic, and then later let the student attain the correct syntax.

Since Parsons problems help explain logic without burdening the student to type the syntax correctly, I would strongly encourage having lines to drag and drop, à la https://js-parsons.github.io/ 

[sensibility]: https://scholar.google.ca/citations?user=A14vyCAAAAAJ&hl=es#d=gs_md_cita-d&u=%2Fcitations%3Fview_op%3Dview_citation%26hl%3Des%26user%3DA14vyCAAAAAJ%26citation_for_view%3DA14vyCAAAAAJ%3A3fE2CSJIrl8C%26tzom%3D360

> * As a way of 'revealing' the solution to arbitrary exercises. Automatically giving the user a shuffled solution for any exercise is easy. I'm just not sure what the user experience should be. We currently allow the user to reveal the solution bit by bit until they see the whole thing. If we simply replaced this entirely with a Parsons problem, what if the user gets really stuck and can't solve that either? If we add a final button to just show the correctly ordered solution, how do we prevent the temptation to just click that immediately? Is there anything from the existing solution reveal method that's worth keeping? Should we give the user a choice between solution reveal methods?

I would avoid this use case for now 😄

**Comment by alexmojaki** [comment]
>  A major stumbling block for novice coders is syntax — getting all the parentheses, colons, and everything in the right order (I should know).

> When teaching about control flow especially, I'd be keen to the solution expressed as Parsons problems first. I have anecdotal experience of students tripping over the colon at the end of if- and for- statements, and unsure about the role of indentation. I think introducing these concepts with Parsons problems makes it easier to focus on logic, and then later let the student attain the correct syntax.

It sounds like you're saying Parsons problems don't help them to learn syntax, so you're expecting the same amount of confusion and frustration either way, but you'd like to push that further down the line to create a positive first impression of coding. Is that right?

Look through the for loops chapter, which is where control flow, indentation, and colons first appear. It starts out heavily focused on syntax and gradually transitions to solving problems. You're suggesting flipping the order, which would mean a complete radical rewrite. I would need to see a proposal to consider this, and would probably need some convincing.

I would be much more interested in having something like sensibility (great paper title btw) in futurecoder so that when students make a syntax error they get automatic suggestions for fixes, thus reducing stumbling and frustration. Do you think you could implement sensibility for Python? That would be an amazing contribution and you're basically the only person who can do it. Plus we could get data of its effectiveness in a real learning setting.

> I would avoid this use case for now

Aww, why?

I've been thinking about it some more. I think that when the user runs out of hints and wants to see a solution, they should be given a parsons problem, along with a button to skip to the usual [gradual solution reveal](https://raw.githubusercontent.com/alexmojaki/futurecoder/master/images/solution.png). Some students might just skip to the solution reveal without attempting the parsons problem, but that's not any worse than what they can do now. Otherwise it gives students one last chance to somewhat solve the problem on their own, which I think is harder to do with the gradual solution reveal.

**Comment by eddieantonio** [comment]
> > A major stumbling block for novice coders is syntax — getting all the parentheses, colons, and everything in the right order (I should know).
> 
> > When teaching about control flow especially, I'd be keen to the solution expressed as Parsons problems first. I have anecdotal experience of students tripping over the colon at the end of if- and for- statements, and unsure about the role of indentation. I think introducing these concepts with Parsons problems makes it easier to focus on logic, and then later let the student attain the correct syntax.
> 
> It sounds like you're saying Parsons problems don't help them to learn syntax, so you're expecting the same amount of confusion and frustration either way, but you'd like to push that further down the line to create a positive first impression of coding. Is that right?

Sort of. I guess there's really two views of teaching coding:

 1. code is learning logic: therefore syntax is just some inconvenient hurdle in the way of the precious computational thinking; or,
 2. code is learning a language: in which case, syntax is _part_ of the lesson

If the existing material assumes the latter, then I suppose it's better to use Parsons problems the way you originally suggested.

> I would be much more interested in having something like sensibility (great paper title btw) in futurecoder so that when students make a syntax error they get automatic suggestions for fixes, thus reducing stumbling and frustration. Do you think you could implement sensibility for Python? That would be an amazing contribution and you're basically the only person who can do it. Plus we could get data of its effectiveness in a real learning setting.

If I had the spare time to do this, I'd love to! As it stands, I can't promise anything :/

> > I would avoid this use case for now
> 
> Aww, why?
> 
> I've been thinking about it some more. I think that when the user runs out of hints and wants to see a solution, they should be given a parsons problem, along with a button to skip to the usual [gradual solution reveal](https://raw.githubusercontent.com/alexmojaki/futurecoder/master/images/solution.png). Some students might just skip to the solution reveal without attempting the parsons problem, but that's not any worse than what they can do now. Otherwise it gives students one last chance to somewhat solve the problem on their own, which I think is harder to do with the gradual solution reveal.

I've given this some thought, and I agree, this is a great idea! I think it would be really nice to see this implemented!

**Comment by alexmojaki** [comment]
OK, I've done a first pass at this in #69 (nice). Please view the deployment there and let me know what you think!

Basically any step can be annotated to allow showing the solution as a parsons problem. This is basically done whenever it doesn't reveal too much information, which is generally when there are at least 5 lines and the task isn't just "slightly modify this given program". In these steps, the parsons problem is shown after the hints, followed by the gradual solution reveal. Otherwise it's just the gradual reveal, or nothing at all if there's no hints.

The overall goal when a student has reached this stage is to find a balance between:

1. Reducing cognitive load. By this time students are furstrated and exhausted and want to move on.
2. Giving students a problem to solve and mentally engage with to improve retention from the exercise.

This balance is very tricky. Some decisions which affect this:

1. When they click one of the two "Show solution" buttons, it just says "Are you sure?" to avoid mistakes, but nothing more. It used to be "Are you sure? You will learn much better if you can solve this yourself" but I've decided that's too forceful. Similarly I considered making some of those buttons red but decided against it.
2. The gradual solution reveal is shown at the same time as the parsons problem. This can make it easy for students to see the solution without revealing the whole thing by combining the two partial solutions. I decided not to hide the parsons problem because at this stage students are especially exhausted and combining the solutions is still a mental effort.
3. Currently it's a 1D Parsons problem, i.e. the indentation is given. I think I want to upgrade it to 2D. Just something to keep in mind, I'm going to defer both the implementation and the discussion for now.

Regarding sensibility/unnaturalcode: if I tried to get it working myself with Python syntax errors, do you foresee any problems? Any guidance you can offer?

**Comment by alexmojaki** [comment]
@spamegg1 check it out

**Comment by alexmojaki** [comment]
@eddieantonio I started looking into sensibility. I see in the source code some signs that Python is supported although it's not mentioned in the README. More importantly I see in the paper that it takes days to train the model on high end GPUs :sweat_smile: You wouldn't happen to have a pretrained Python model lying around would you?

**Comment by eddieantonio** [comment]
> @eddieantonio I started looking into sensibility. I see in the source code some signs that Python is supported although it's not mentioned in the README. More importantly I see in the paper that it takes days to train the model on high end GPUs 😅 You wouldn't happen to have a pretrained Python model lying around would you?

I do _not_ have pre-trained models for Python, unfortunately :/

Frankly, I'd suggest creating lightweight **bi-directional** n-gram models instead of LSTMs (or any kind of neural network), since they're lightweight, easy to "train" (estimate). We did not evaluate using bi-directional n-grams, but I assume they would provide comparable effectiveness to the LSTMs at a far lower compuational cost.

**Comment by alexmojaki** [comment]
> I do _not_ have pre-trained models for Python, unfortunately :/

:sob: 

Got some spare GPU compute? :eyes:

Let's continue this discussion in #70.

Have you had a chance to play with Parsons problem solutions? Any thoughts on what I wrote above?

**Comment by alexmojaki** [comment]
@eddieantonio @spamegg1 please check out #73

**Comment by alexmojaki** [comment]
Continued in #81

**Comment by alexmojaki** [comment]
Also see #95
